### PR TITLE
Kubectl exec support resource/name format

### LINKED ISSUE #72104: kubectl exec does not support resource/name format
Unfortunately, not even `pod/pod-name` is supported which makes using `kubectl exec` in scripts very annoying.

@kubernetes/sig-cli-feature-requests

### PR DESCRIPTION
**Pull request by ahmadnurus**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature
> /kind flake

**What this PR does / why we need it**:

Add support kubectl to execute container from pod selected by resource/name.

**Which issue(s) this PR fixes**:

Fixes #72104 

**Special notes for your reviewer**:

cc @soltysh 

like what port-forward does, i'm using `AttachablePodForObject` to select a pod using resource name. but I resolve the pod from the resource name in the `Run()` method. because  if it's on the `Complete()` method this makes panic when retrieving the object from `builder.Do().Object()` when do a test. this can be compiled correctly and running properly. it's just that when testing it causes panic.

**Does this PR introduce a user-facing change?**:

```release-note
kubectl exec now allows using resource name (e.g., deployment/mydeployment) to select a matching pod.
kubectl exec now allows using --pod-running-timeout flag to wait till at least one pod is running.
```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
/priority important-longterm

**Review by soltysh — CHANGES_REQUESTED** [review]
I left you some comments, additionally I'd like to see a test (can be test-cmd) verifying that both the old way of specifying just pod name and the new with resource/name work equally well.

File: pkg/kubectl/cmd/exec/exec.go [hunk 1]
```diff
@@ -84,8 +97,7 @@ func NewCmdExec(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.C
 			cmdutil.CheckErr(options.Run())
 		},
 	}
-	cmd.Flags().StringVarP(&options.PodName, "pod", "p", options.PodName, "Pod name")
```
[the remark this answers is not in the document]
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: well, I will bring back the -p flag

**Comment by prksu** [comment]
@soltysh updated, and i make a test-cmd script to verifying both ways work as well. PTAL

File: test/cmd/exec.sh [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+run_kubectl_exec_pod_tests() {
+	set -o nounset
```
Comment (soltysh) [inline test/cmd/exec.sh]: Can you please fix formatting in this file?

File: test/cmd/legacy-script.sh [hunk 3]
```diff
@@ -484,6 +485,15 @@ runTests() {
     record_command run_kubectl_old_print_tests
   fi
 
+  ################
+  # Kubectl exec #
+  ################
+
+  if kube::test::if_supports_resource "${pods}" ; then
+    record_command run_kubectl_exec_pod_tests
+    record_command run_kubectl_exec_resource_name_tests
```
Comment (soltysh) [inline test/cmd/legacy-script.sh]: Not quite, in this test you require pods, deployments, replicaset and configmaps too.

**Review by soltysh — CHANGES_REQUESTED** [review]
/ok-to-test

Som more comments

**Comment by prksu** [comment]
@soltysh this PR breaking `kubectl cp` command. i have add one commit as my solution. will be rebase if you aprove this solution. or you have another better solution than this?

**Review by soltysh — CHANGES_REQUESTED** [review]
Maybe then just move the test to e2e, https://github.com/kubernetes/kubernetes/blob/395e4c05bad2869be91e89009d47d43a90089b56/test/e2e/kubectl/kubectl.go#L367-L405 is a good place

**Comment by prksu** [comment]
ok, i will move the test to e2e.

what about the breaking `kubectl cp` command? do you agree with this fa608fd251dc8c676fb6eef240c6a83498e90219 ?

**Comment by soltysh** [comment]
> what about the breaking `kubectl cp` command? do you agree with this [fa608fd](https://github.com/kubernetes/kubernetes/commit/fa608fd251dc8c676fb6eef240c6a83498e90219) ?

The changes are reasonable, but we can't break `kubectl cp` in any way.

**Comment by kargakis** [comment]
Why is kubectl cp broken by this? @prksu @soltysh would love to see this merged.

**Comment by prksu** [comment]
@kargakis on my last commit it was no longer broken. it's just that this needs a some e2e test and I'll do it soon. sorry for the delay

**Comment by kargakis** [comment]
@prksu any updates? If you are not able to move this PR forward, @droslean may be able to pick it up

**Comment by kargakis** [comment]
/lgtm
/retest 

thanks @prksu !

**Comment by prksu** [comment]
@soltysh @kargakis, i have added e2e test to kubect exec using resource/name. i just added one e2e test in `Simple Pod` block as you show @soltysh. not sure if I should add e2e test for kubectl exec independently because I see there is nothing.

and sorry for delay.

File: pkg/kubectl/cmd/exec/exec.go [hunk 4]
```diff
@@ -84,6 +98,7 @@ func NewCmdExec(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.C
 			cmdutil.CheckErr(options.Run())
 		},
 	}
+	cmdutil.AddPodRunningTimeoutFlag(cmd, defaultPodExecTimeout)
 	cmd.Flags().StringVarP(&options.PodName, "pod", "p", options.PodName, "Pod name")
 	cmd.Flags().MarkDeprecated("pod", "This flag is deprecated and will be removed in future. Use exec POD_NAME instead.")
```
Comment (kargakis) [inline pkg/kubectl/cmd/exec/exec.go]: Can you open a separate issue to remove this flag? I think it's marked deprecated for a couple of releases now so we should be able to remove it. It would simplify a lot of the code.

**Review by 0xmichalis — APPROVED** [review]
/lgtm

@soltysh needs an approval

File: pkg/kubectl/cmd/exec/exec.go (hunk 4 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: do you mean `--pod` flag?
`--pod` flag was deprecated in July last year https://github.com/kubernetes/kubernetes/commit/888119a47cd and I think this should stay around until July next month. of course I am willing to open the issue to delete this flag in time

Comment (soltysh) [reply pkg/kubectl/cmd/exec/exec.go]: Yeah, I complained before, but now 1.15 is a good time to remove the flag. Open a separate PR doing this, with the release date planned sometime in June we're good to remove it. Ping me in the PR please :smile:

File: pkg/kubectl/cmd/exec/exec.go [hunk 5]
```diff
@@ -266,11 +294,34 @@ func (o *StreamOptions) SetupTTY() term.TTY {
...
+	if len(p.PodName) != 0 {
+		p.Pod, err = p.PodClient.Pods(p.Namespace).Get(p.PodName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+	} else {
+		builder := p.Builder().
+			WithScheme(scheme.Scheme, scheme.Scheme.PrioritizedVersionsAllGroups()...).
+			NamespaceParam(p.Namespace).DefaultNamespace().
+			ResourceNames("pods", p.ResourceName)
```
Comment (soltysh) [inline pkg/kubectl/cmd/exec/exec.go]: This is wrong, you need to pass the actual type and name here, compare with https://github.com/kubernetes/kubernetes/blob/fd98e7abf042251bb094a5535d67f53d36021246/pkg/kubectl/cmd/attach/attach.go#L230. Also I'd encourage you to switch from this conditions to a code similar in the linked attach.

File: pkg/kubectl/cmd/exec/exec.go [hunk 6]
```diff
@@ -330,3 +381,12 @@ func (p *ExecOptions) Run() error {
 
 	return nil
 }
+
+func (p *ExecOptions) findExecutablePod(obj runtime.Object) (*corev1.Pod, error) {
```
Comment (soltysh) [inline pkg/kubectl/cmd/exec/exec.go]: I don't see any reason you need this method, for testing you can inject `ExecutablePodFn` into `ExecOptions`.

**Review by soltysh — CHANGES_REQUESTED** [review]
/lgtm cancel

File: pkg/kubectl/cmd/exec/exec.go (hunk 4 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: OK, I will open a new PR related to that :sweat_smile:

File: pkg/kubectl/cmd/exec/exec.go (hunk 5 above)
Comment (kargakis) [reply pkg/kubectl/cmd/exec/exec.go]: It's not wrong, it's the right way to handle single arguments while maintaining backwards compatibility with the existing format. With your suggestion, exec will now support spaces between resource types and names.

Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: Yes, if I apply something like this.

```
builder := p.Builder().
	WithScheme(scheme.Scheme, scheme.Scheme.PrioritizedVersionsAllGroups()...).
	NamespaceParam(p.Namespace).DefaultNamespace()

if len(p.PodName) != 0 {
	builder.ResourceNames("pods", p.PodName)
} else {
	builder.ResourceNames("pods", p.ResourceName)
}
```

it's cause a panic on `kubectl cp` command. because in `kubectl cp` does not fill the `Builder` in `ExecOptions` struct https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/cp/cp.go#L506-L526

Comment (kargakis) [reply pkg/kubectl/cmd/exec/exec.go]: I think what @soltysh meant is to end up with is something like the following:
```go
	if len(p.PodName) != 0 {
		p.Pod, err = p.PodClient.Pods(p.Namespace).Get(p.PodName, metav1.GetOptions{})
		if err != nil {
			return err
		}
	} else {
		b := p.Builder().
			WithScheme(scheme.Scheme, scheme.Scheme.PrioritizedVersionsAllGroups()...).
			NamespaceParam(p.Namespace).DefaultNamespace()

		switch len(o.Resources) {
		case 1:
			b.ResourceNames("pods", o.Resources[0])
		case 2:
			b.ResourceNames(o.Resources[0], o.Resources[1])
}
,,,

Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: hmmm, I misunderstood that 😄 

so like @kargakis say, whether `kubectl exec` will now support spaces between resource types and names?

Comment (kargakis) [reply pkg/kubectl/cmd/exec/exec.go]: I think it's an acceptable thing to support but to be fair the original issue was about adding `type/name` support. I'd prefer to get this merged, and add `type name` support in a follow-up (also needs to be added in other commands such as `kubectl logs`)

File: pkg/kubectl/cmd/exec/exec.go (hunk 6 above)
Comment (kargakis) [reply pkg/kubectl/cmd/exec/exec.go]: @prksu please address this comment, thanks

File: pkg/kubectl/cmd/exec/exec.go (hunk 5 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: if kubectl exec support spaces between resource types and names like kubectl attach. how we can mapping between kubectl exec arguments and command with argument? like this `POD COMMAND [ARG]`, `TYPE/NAME COMMAND [ARG]` and `TYPE NAME COMMAND` :sweat_smile:

File: pkg/kubectl/cmd/exec/exec.go (hunk 6 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: yes soltysh is right about that. `findExecutablePod` method is not necessary. I just updated my commit.

**Comment by kargakis** [comment]
@soltysh ptal, needs an approval

**Comment by prksu** [comment]
ping @soltysh

File: pkg/kubectl/cmd/exec/exec.go (hunk 5 above)
Comment (soltysh) [reply pkg/kubectl/cmd/exec/exec.go]: I'd rather get it fixed in this PR while this is being modified. I know from experience that followups might slip away and that won't happen, so while at it please straighten this one up.

**Review by soltysh — CHANGES_REQUESTED** [review]
That comment of mine wrt resource builder being use is still not addressed. I won't let this in until it gets properly fixed.
/hold

File: pkg/kubectl/cmd/exec/exec.go (hunk 5 above)
Comment (kargakis) [reply pkg/kubectl/cmd/exec/exec.go]: @soltysh maybe @prksu's comment wasn't clear enough? Your suggestion about supporting spaces between resource types and names is actually a breaking one in `kubectl exec`.

Currently the following `kubectl exec` invocation is valid:
```console
$ kubectl exec weave-net-6xblq date
Defaulting container name to weave.
Use 'kubectl describe pod/weave-net-6xblq -n kube-system' to see all of the containers in this pod.
Mon Apr  8 20:30:45 UTC 2019
```
`date` is executed in the container (without having to pass `--` in order to separate between the kubernetes resource and the command). With your suggestion `weave-net-6xblq` would be parsed as a resource type and `date` would be parsed as a resource name, right?

Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: @kargakis maybe what @soltysh wants is just passing the actual type and name in resource builder. without supporting spaces between resource types and names. it's right @soltysh ?

Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: i just updated. PTAL

**Comment by kargakis** [comment]
@soltysh ptal otherwise assign to someone else with more bandwidth to move this forward, this PR is sitting in for two months now ...

**Comment by prksu** [comment]
ping @soltysh

File: pkg/kubectl/cmd/exec/exec.go (hunk 4 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: @soltysh I have open separate PR to remove this flag.

File: pkg/kubectl/cmd/exec/exec.go [hunk 7]
```diff
@@ -159,32 +181,39 @@ func (p *ExecOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, argsIn []s
 		}
 		p.Command = argsIn
 	} else {
-		p.PodName = argsIn[0]
+		p.Resources = strings.Split(argsIn[0], "/")
```
Comment (kargakis) [inline pkg/kubectl/cmd/exec/exec.go]: I don't think this is the right approach, the resource builder should already do the parsing for us. See how this is implemented in `kubectl logs`.

File: pkg/kubectl/cmd/exec/exec.go [hunk 8]
```diff
@@ -266,11 +292,40 @@ func (o *StreamOptions) SetupTTY() term.TTY {
...
+		p.Pod, err = p.PodClient.Pods(p.Namespace).Get(p.PodName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+	} else {
+		builder := p.Builder().
+			WithScheme(scheme.Scheme, scheme.Scheme.PrioritizedVersionsAllGroups()...).
+			NamespaceParam(p.Namespace).DefaultNamespace()
+
+		switch len(p.Resources) {
```
Comment (kargakis) [inline pkg/kubectl/cmd/exec/exec.go]: The whole switch statement is redundant, see how [ResourceNames](https://github.com/kubernetes/cli-runtime/blob/cdb522bcb34cfd041c0f6473be10c804e45eb0b3/pkg/resource/builder.go#L398) is implemented. The first argument should be "pods", ie. the default resource, and the second argument should be `argsIn[0]` (without splitting it).

File: pkg/kubectl/cmd/exec/exec.go (hunk 7 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: @kargakis

Yeah, before i see how kubctl logs and kubectl port-forward implemented the resource builder.

I did this to addressing soltysh comment about passing the actual type and name.

I'm a little confused here :confused:

File: pkg/kubectl/cmd/exec/exec.go (hunk 8 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: Same with the answer to your comment above.a

I did this to addressing soltysh comment about passing the actual type and name.

File: pkg/kubectl/cmd/exec/exec.go (hunk 7 above)
Comment (kargakis) [reply pkg/kubectl/cmd/exec/exec.go]: This is not how `kubectl attach` (the example brought up by @soltysh ) parses arguments. In @soltysh's example the first argument in `ResourceNames` is the first argument passed in `kubectl attach` and the second argument in `ResourceNames` is the second arg in `kubectl attach` (see [here](https://github.com/kubernetes/kubernetes/blob/cee320a809e10d9147b634e4e2b05417c79a490b/pkg/kubectl/cmd/attach/attach.go#L106), then [here](https://github.com/kubernetes/kubernetes/blob/cee320a809e10d9147b634e4e2b05417c79a490b/pkg/kubectl/cmd/attach/attach.go#L179), then [here](https://github.com/kubernetes/kubernetes/blob/cee320a809e10d9147b634e4e2b05417c79a490b/pkg/kubectl/cmd/attach/attach.go#L226-L231)). What you have done here, is take the first argument in `kubectl exec`, split it, and pass those args in `ResourceNames`. This is not what @soltysh asked, but to his request I am not sure `kubectl exec` can support the `TYPE NAME` format because the second argument passed to `kubectl exec` is already parsed as a command to be executed in the pod (see [here](https://github.com/kubernetes/kubernetes/blob/cee320a809e10d9147b634e4e2b05417c79a490b/pkg/kubectl/cmd/exec/exec.go#L163)).

File: pkg/kubectl/cmd/exec/exec.go (hunk 8 above)
Comment (kargakis) [reply pkg/kubectl/cmd/exec/exec.go]: See my reply [above](https://github.com/kubernetes/kubernetes/pull/73664/files#r276767421). All in all, I think what you had initially is the right approach.

File: pkg/kubectl/cmd/exec/exec.go (hunk 7 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: @kargakis I see. 

I am not sure too what @soltysh is asking for is `kubectl exec` can support `TYPE NAME`. I have questioned this in my previous reply above. if what @soltysh requesting is `kubectl exec` can support `TYPE NAME`, I have no idea how implement this.

File: pkg/kubectl/cmd/exec/exec.go (hunk 8 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: I think so.

File: pkg/kubectl/cmd/exec/exec.go [hunk 9]
```diff
@@ -40,27 +44,37 @@ import (
 
 var (
 	execExample = templates.Examples(i18n.T(`
-		# Get output from running 'date' from pod 123456-7890, using the first container by default
-		kubectl exec 123456-7890 date
+		# Get output from running 'date' command from pod mypod,
+		# using the first container by default
```
Comment (soltysh) [inline pkg/kubectl/cmd/exec/exec.go]: Good change :+1: but please the formatting (iow. the description in a single line) like before. I don't want to end up with too long help like in `kubectl run` case.

File: pkg/kubectl/cmd/exec/exec.go [hunk 10]
```diff
@@ -40,27 +44,37 @@ import (
...
+		# Get output from running 'date' command from first pod of the deployment named mydeployment,
+		# using the first container by default
+		kubectl exec deploy/mydeployment date
 
-		# Switch to raw terminal mode, sends stdin to 'bash' in ruby-container from pod 123456-7890
+		# Get output from running 'date' command from first pod of the service named myservice,
+		# using the first container by default
+		kubectl exec svc/myservice date
+
+		# Get output from running 'date' command in ruby-container from pod mypod
```
Comment (soltysh) [inline pkg/kubectl/cmd/exec/exec.go]: Put this example next to the first one, to keep the consistency as before.

File: pkg/kubectl/cmd/exec/exec.go [hunk 11]
```diff
@@ -40,27 +44,37 @@ import (
...
 	execExample = templates.Examples(i18n.T(`
-		# Get output from running 'date' from pod 123456-7890, using the first container by default
-		kubectl exec 123456-7890 date
+		# Get output from running 'date' command from pod mypod,
+		# using the first container by default
+		kubectl exec mypod date
 
-		# Get output from running 'date' in ruby-container from pod 123456-7890
-		kubectl exec 123456-7890 -c ruby-container date
+		# Get output from running 'date' command from first pod of the deployment named mydeployment,
```
Comment (soltysh) [inline pkg/kubectl/cmd/exec/exec.go]: **the** first

Comment (soltysh) [reply pkg/kubectl/cmd/exec/exec.go]: deployment ~~named~~ mydeployment

File: pkg/kubectl/cmd/exec/exec.go [hunk 12]
```diff
@@ -40,27 +44,37 @@ import (
...
+		# using the first container by default
+		kubectl exec deploy/mydeployment date
 
-		# Switch to raw terminal mode, sends stdin to 'bash' in ruby-container from pod 123456-7890
+		# Get output from running 'date' command from first pod of the service named myservice,
```
Comment (soltysh) [inline pkg/kubectl/cmd/exec/exec.go]: **the** first

Comment (soltysh) [reply pkg/kubectl/cmd/exec/exec.go]: service ~~named~~ myservice

File: pkg/kubectl/cmd/exec/exec.go (hunk 7 above)
Comment (soltysh) [reply pkg/kubectl/cmd/exec/exec.go]: @kargakis is right, I want to see these changes here: https://github.com/kubernetes/kubernetes/blob/cee320a809e10d9147b634e4e2b05417c79a490b/pkg/kubectl/cmd/attach/attach.go#L179 in `Complete` https://github.com/kubernetes/kubernetes/blob/cee320a809e10d9147b634e4e2b05417c79a490b/pkg/kubectl/cmd/attach/attach.go#L206-L214 in `Validate` and https://github.com/kubernetes/kubernetes/blob/cee320a809e10d9147b634e4e2b05417c79a490b/pkg/kubectl/cmd/attach/attach.go#L226-L231 in `Run`.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/exec/exec.go (hunk 7 above)
Comment (kargakis) [reply pkg/kubectl/cmd/exec/exec.go]: @soltysh isn't the following suggestion incompatible with `kubectl exec` as it stands?
```go
 case 2: 
b.ResourceNames(o.Resources[0], o.Resources[1]) 
```
`kubectl exec` already parses the second argument (`o.Resources[1]`) as a command to be executed inside the pod (see [here](https://github.com/kubernetes/kubernetes/blob/cee320a809e10d9147b634e4e2b05417c79a490b/pkg/kubectl/cmd/exec/exec.go#L163))

File: pkg/kubectl/cmd/exec/exec.go (hunk 9 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: updated

File: pkg/kubectl/cmd/exec/exec.go (hunk 10 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: updated

File: pkg/kubectl/cmd/exec/exec.go (hunk 11 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: updated

File: pkg/kubectl/cmd/exec/exec.go (hunk 12 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: updated, thanks @soltysh

File: pkg/kubectl/cmd/exec/exec.go (hunk 7 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: @soltysh can you point me about this one? 

how we can separate `TYPE` and `NAME` with whitespace if second argument in `kubectl exec` is a `COMMAND` that will be executed inside the pod? like my previous question how we can dealing with something like this?

| CMD          | ARG1      | ARG2    |
|--------------|-----------|---------|
| kubectl exec | POD       | COMMAND |
| kubectl exec | TYPE/NAME | COMMAND |
| kubectl exec | TYPE      | NAME    |

different with `kubectl attach` he don't have `COMMAND` as second argument. so it's fine `kubectl attach` meet something like this.

| CMD          | ARG1      | ARG2    |
|--------------|-----------|---------|
| kubectl attach | POD       | |
| kubectl attach | TYPE/NAME | |
| kubectl attach | TYPE      | NAME    |

like @kargakis say, i think my previous approach was quite right.

Comment (kargakis) [reply pkg/kubectl/cmd/exec/exec.go]: @prksu sorry for going back and forth, please go ahead and revert to your previous PR state

Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: @kargakis i have reverted this to my previous approach.

File: test/cmd/exec.sh [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+  # Clean up
+  kubectl delete pods test-pod
+
+  set +o nounset
+  set +o errexit
+}
+
+run_kubectl_exec_resource_name_tests() {
+  set +o nounset
+  set +o errexit
```
Comment (kargakis) [inline test/cmd/exec.sh]: Shouldn't this be -o?

Comment (kargakis) [reply test/cmd/exec.sh]: Or you need to avoid exiting when you test failure scenarios?

File: pkg/kubectl/cmd/exec/exec.go [hunk 14]
```diff
@@ -266,11 +288,35 @@ func (o *StreamOptions) SetupTTY() term.TTY {
...
+		p.Pod, err = p.PodClient.Pods(p.Namespace).Get(p.PodName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+	} else {
+		builder := p.Builder().
+			WithScheme(scheme.Scheme, scheme.Scheme.PrioritizedVersionsAllGroups()...).
+			NamespaceParam(p.Namespace).DefaultNamespace()
+
+		builder.ResourceNames("pods", p.ResourceName)
```
Comment (kargakis) [inline pkg/kubectl/cmd/exec/exec.go]: nit: chain this with the rest of the builder methods you execute above

File: test/cmd/exec.sh (hunk 13 above)
Comment (prksu) [reply test/cmd/exec.sh]: @kargakis I just following other test cmd

**Comment by kargakis** [comment]
/lgtm
/hold cancel

@kubernetes/sig-cli-pr-reviews needs an approval

File: pkg/kubectl/cmd/exec/exec.go (hunk 14 above)
Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: @kargakis i will update it soon

Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: updated
thanks @kargakis

File: pkg/kubectl/cmd/exec/exec.go [hunk 15]
```diff
@@ -159,32 +177,39 @@ func (p *ExecOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, argsIn []s
 		}
 		p.Command = argsIn
 	} else {
-		p.PodName = argsIn[0]
+		p.ResourceName = argsIn[0]
 		p.Command = argsIn[1:]
-		if len(p.Command) < 1 {
```
Comment (kargakis) [inline pkg/kubectl/cmd/exec/exec.go]: Why is this part removed?

Comment (prksu) [reply pkg/kubectl/cmd/exec/exec.go]: it is will validate in `Validate` method in this line https://github.com/kubernetes/kubernetes/blob/0c39d7d3806f1a45148f159b8480ab86df9f23ff/pkg/kubectl/cmd/exec/exec.go#L228-L230

i think this part is not necessary, so i leave it validated on the `Validate` method

Comment (kargakis) [reply pkg/kubectl/cmd/exec/exec.go]: Makes sense, thanks!

**Review by janetkuo — APPROVED** [review]

**Comment by kfox1111** [comment]
This is great. Thanks. :)

Any chance that the non interactive mode can run the command on all matching pods rather then just the first one?

**Comment by prksu** [comment]
> This is great. Thanks. :)
> 
> Any chance that the non interactive mode can run the command on all matching pods rather then just the first one?

you mean bulk exec? unfortunately i think `kubectl exec` doesn't support that. you can open a new issue with that topic with the feature request format.

or you can implement this as `kubectl plugin`. like https://github.com/txn2/kubefwd
